Publisher

rexhart opened the pull request:
# Changelog

## [Unreleased]

### Added
- New CSV publisher tool for sending transaction data to SQS  'cmd/csv_publisher/publisher.go'
  - Processes CSV files in configurable batch sizes  
  - Maintains state between runs to avoid duplicate processing  
  - Supports resuming from last processed position  
  - Handles concurrent batch processing with goroutines  
- SQS messaging functionality in `internal/messaging/sqs.go`  
  - Added `SQSHandler` for sending and receiving transactions  
  - Implemented JSON serialization/deserialization of transactions  
  - Environment variable support for queue URL and CSV file path  
- AWS SQS client dependency in `go.mod`  

### Changed
- Updated `internal/config/config.go` to:  
  - Look for `.env` file in multiple locations  
  - Add SQS configuration settings  
  - Improve environment variable handling  
- Modified AWS Lambda batch size in `deployments/template.yaml` from `10` to `1`  
  - Ensures individual transaction processing for better error handling  

### Fixed
- Resolved package naming conflicts between AWS SDK config and internal config  

### Notes for Reviewers
- This PR adds a CSV publisher tool that reads transaction data from CSV files and publishes it to an SQS queue for processing.  
- The publisher maintains state between runs to avoid duplicate processing and supports resuming from the last processed position.  
- It also adds SQS messaging functionality to the internal messaging package.  
- The changes are backward compatible and should not affect existing functionality.  
- The CSV publisher is a new standalone tool that can be used to load test data into the system.

rexhart committed:
for joe to see

rexhart committed:
updated sqs link config.go

rexhart committed:
fixed some imports in main.go etc

rexhart committed:
Merge branch 'main' into rex

rexhart committed:
hi

rexhart committed:
Merge branch 'main' into rex

rexhart committed:
some publisher stuff

rexhart committed:
Merge branch 'main' into rex

rexhart committed:
some more merging

rexhart committed:
removed main.go

rexhart committed:
some publisher things

rexhart committed:
fixed config.go InitializeConfig

wenjielee11 commented on internal/config/config.go:
```diff
@@ -41,6 +48,11 @@ var SNSMessengerConfig = &struct {
+var SQSConfig = &struct {
+	QueueURL string
+}{}
+
```
Where is SQSConfig Initialized? You might want to consider doing that in LoadAWSConfig, or through some other method since it might conflict with our consumer stack

rexhart committed:
config.go sqs init

rexhart committed:
publisher updates

wenjielee11 commented on deployments/template.yaml:
```diff
@@ -31,6 +31,11 @@ Parameters:
     Description: AWS region where resources should be deployed
     Default: us-east-1
 
+  QueueUrl:
```
I dont see this parameter being used anywhere in the current template. Unused code from when testing maybe? You might want to consider making a separate template file for deploying and testing ur publisher if you'd like. Or maybe just a script or makefile to run that spins up the SQS queue and tears it down when needed. Then a publisher script(which you already have) to publish X amount of transactions. Many ways to go about this. Preferably minimizing the use of AWS console to setup stuff and keep the setup process coded or scripted

wenjielee11 commented on internal/config/config.go:
```diff
@@ -14,9 +14,16 @@ import (
+				log.Println("Warning: No .env file found or failed to load.")
+			}
+		}
 	}
```
I think theres a better way to do this other than the if statements:
https://stackoverflow.com/questions/68346410/how-to-use-dynamic-location-for-godotenv-load-env-file#:~:text=16-,Following,-the%20suggestion%20of

wenjielee11 commented on internal/messaging/sqs.go:
```diff
@@ -1 +1,69 @@
+	var transactions []*models.Transaction
+	for _, msg := range output.Messages {
+		var transaction models.Transaction
+		if err := json.Unmarshal([]byte(*msg.Body), &transaction); err != nil {
```
For the marshaling and unmarshaling across your code, you can consider trying out the helper functions under the transaction model to see if it works or not. You can also call the validate function before publishing the transaction to check for integrity. Though, if it works theres no need to change it

wenjielee11 commented on test/publisher_test.go:
```diff
@@ -0,0 +1,269 @@
+		Email:             "test@example.com",
+		TransactionStatus: "PENDING",
+	}
+
```
Your publisher test should do a model validation first before sending. Useful for when we change the transaction model in the future(when using another dataset), you will get notified that this part has to be fixed to follow the new change. So ideally, the moment models.Transaction gets updated, all tests should fail and prompt us to fix our code to adapt to the new model.

wenjielee11 reviewed (commented):
Awesome work!

rexhart committed:
Merge branch 'main' into rex

rexhart committed:
publisher updates + extra debug statements

rexhart committed:
pr review updates

rexhart commented:
New CSV publisher tool for sending transaction data to SQS 'cmd/csv_publisher/publisher.go'
Processes CSV files in configurable batch sizes
Maintains state between runs to avoid duplicate processing
Supports resuming from last processed position
Handles concurrent batch processing with goroutines
SQS messaging functionality in internal/messaging/sqs.go
Added SQSHandler for sending and receiving transactions
Implemented JSON serialization/deserialization of transactions
Environment variable support for queue URL and CSV file path
AWS SQS client dependency in go.mod
Changed
Updated internal/config/config.go to:
Look for .env file in multiple locations
Add SQS configuration settings
Improve environment variable handling
Modified AWS Lambda batch size in deployments/template.yaml from 10 to 1
Ensures individual transaction processing for better error handling
